File: python/dynamic_programming/triangle.py
class Solution:
    # @param triangle, a list of lists of integers
    # @return an integer
    def minimumTotal(self, triangle):
        opt = [0 for i in xrange(len(triangle[-1]))]
        temp = [0 for i in xrange(len(triangle[-1]))]
        
        for i in xrange(len(triangle)):
            for j in xrange(len(triangle[i])):
                if j == 0:
                    temp[j] += triangle[i][j]
                elif j == len(triangle[i])-1:
                    temp[j] = opt[j-1] + triangle[i][j] 
                else:
                    temp[j] = min(opt[j], opt[j-1]) + triangle[i][j]
            opt = temp[:]
        
        return min(opt)

if __name__ == "__main__":
    solution = Solution()

    #print solution.minimumTotal([[-10]])
    print solution.minimumTotal([[2], [3,4], [6,5,7], [4,1,8,3]])

File: java/Triangle.java
public class Solution {
    public int minimumTotal(List<List<Integer>> triangle) {
        int[] opt = new int[triangle.get(triangle.size()-1).size()];
        Arrays.fill(opt, 0);
        int[] opt_temp = Arrays.copyOfRange(opt, 0, opt.length);
        for (int i = 0; i < triangle.size(); i++) {
            for (int j = 0; j < triangle.get(i).size(); j++) {
                int left = opt[j] + triangle.get(i).get(j);
                int right = 0;
                if (j > 0) {
                    right = opt[j-1] + triangle.get(i).get(j);
                }
                if (j == 0) {
                    opt_temp[j] = left;
                } else if (j == triangle.get(i).size()-1) {
                    opt_temp[j] = right;
                } else {
                    opt_temp[j] = Math.min(left, right);
                }
            }
            opt = Arrays.copyOfRange(opt_temp, 0, opt_temp.length);
        }
        int result = 0;
        for (int i = 0; i < opt.length; i++) {
            if (i == 0 || opt[i] < result) {
                result = opt[i];
            }
        }
        return result;
    }
}
